protocol/nats_jetstream/v3/protocol.go

```diff
@@ -0,0 +1,290 @@
+		// close connection if an error occurs
+		if errConnection != nil {
+			p.Close(ctx)
+		}
+	}()
+
+	// if a URL was given create the nats connection
+	if p.conn == nil {
+		p.conn, errConnection = nats.Connect(p.url, p.natsOpts...)
+		if errConnection != nil {
+			return nil, errConnection
+		}
+	}
+
+	if p.jetStream, errConnection = jetstream.New(p.conn, p.jetSteamOpts...); errConnection != nil {
+		return nil, errConnection
+	}
+
+	return p, nil
+}
+
+// Send sends messages. Send implements Sender.Sender
+func (p *Protocol) Send(ctx context.Context, in binding.Message, transformers ...binding.Transformer) (err error) {
+	if p.sendSubject == "" {
+		return ErrNoSendSubject
+	}
+
+	defer func() {
+		if err2 := in.Finish(err); err2 != nil {
+			if err == nil {
+				err = err2
+			} else {
+				err = fmt.Errorf("failed to call in.Finish() when error already occurred: %s: %w", err2.Error(), err)
+			}
+		}
+	}()
+
+	if _, err = p.jetStream.StreamNameBySubject(ctx, p.sendSubject); err != nil {
+		return err
+	}
+
+	writer := new(bytes.Buffer)
+	header, err := WriteMsg(ctx, in, writer, transformers...)
+	if err != nil {
+		return err
+	}
+
+	natsMsg := &nats.Msg{
+		Subject: p.sendSubject,
+		Data:    writer.Bytes(),
+		Header:  header,
+	}
+
+	_, err = p.jetStream.PublishMsg(ctx, natsMsg, p.publishOpts...)
+
+	return err
+}
+
+// OpenInbound implements Opener.OpenInbound
+func (p *Protocol) OpenInbound(ctx context.Context) error {
+	p.subMtx.Lock()
+	defer p.subMtx.Unlock()
+
+	var consumeContext jetstream.ConsumeContext
+	var err error
+	if err = p.createJetstreamConsumer(ctx); err != nil {
+		return err
+	}
+	if consumeContext, err = p.jetstreamConsumer.Consume(p.MsgHandler, p.pullConsumeOpts...); err != nil {
+		return err
+	}
+
+	// Wait until external or internal context done
+	select {
+	case <-ctx.Done():
+	case <-p.internalClose:
+	}
+
+	// Finish to consume messages in the queue and close the subscription
+	if consumeContext != nil {
+		consumeContext.Drain()
+	}
+	return nil
+}
+
+// Receive implements Receiver.Receive.
+func (p *Protocol) Receive(ctx context.Context) (binding.Message, error) {
+	select {
+	case msgErr, ok := <-p.incoming:
+		if !ok {
+			return nil, io.EOF
+		}
+		return msgErr.msg, nil
+	case <-ctx.Done():
+		return nil, io.EOF
+	}
+}
+
+// MsgHandler implements nats.MsgHandler and publishes messages onto our internal incoming channel to be delivered
+// via r.Receive(ctx)
+func (p *Protocol) MsgHandler(msg jetstream.Msg) {
+	p.incoming <- msgErr{msg: NewMessage(msg)}
+}
+
+// Close must be called after use to releases internal resources and close the underlying connection
+func (p *Protocol) Close(ctx context.Context) error {
+	// Before closing, let's be sure OpenInbound completes
+	// We send a signal to close and then we lock on subMtx in order
+	// to wait OpenInbound to finish draining the queue
+	p.internalClose <- struct{}{}
+	p.subMtx.Lock()
+	defer p.subMtx.Unlock()
+
+	if p.connOwned && p.conn != nil {
```

Comment:
Do we still want to track owned connections? Currently the comment makes it sound that even if I specify a custom connection, conn would be closed (which would not be the case). I feel we can remove connOwned checks and handle connection closing for both situations

Reply:
The v2 version had this concept.  The idea is that if a connection was passed in, we would use it and then exit without closing the connection (since it was opened before use). 

With a URL, we create a  connection and then close it when Close() is called:
`		p.conn.Close()`

I am not sure if users would complain if we closed their "passed in connection".

Reply:
> I am not sure if users would complain if we closed their "passed in connection".

Well, this is what I was referring to earlier: users are in full control (and it's actually what your doc string says for `Close()`). So users can decide when they underlying connection (either ours or their provided) will be closed.

Reply:
I changed the comments for Close() and added a few new comments.  I removed the `connOwned` member and only use the `url` member.